lib/decoder_impl.cc

```diff
@@ -19,804 +19,1021 @@
+            (void) slide;
 
-    float decoder_impl::detect_upchirp(const gr_complex *samples, uint32_t window, uint32_t slide, int32_t* index) {
-        float samples_ifreq[window];
+            bool found_change = false;
+            uint32_t max = 0u, min;
+            const uint32_t coeff = 16u;
 
-        instantaneous_frequency(samples, samples_ifreq, window);
-        return sliding_norm_cross_correlate(samples_ifreq, &d_upchirp_ifreq[0], window, slide, index);
-    }
+            for (uint32_t i = 0u; i < window - coeff - 1u; i += coeff / 2u) {
+                if (samples[i] - samples[i + coeff]  > 0.2f) { // Goes down
+                    max = i;
+                    found_change = true;
+                    break;
+                }
+            }
 
-    unsigned int decoder_impl::get_shift_fft(gr_complex* samples) {
-        float fft_mag[d_number_of_bins];
-        gr_complex mult_hf[d_samples_per_symbol];
+            if (!found_change) {
+                return 0.0f;
+            }
 
-        #ifdef CFO_CORRECT
-            determine_cfo(&samples[0]);
-            d_debug << "CFO: " << d_cfo_estimation << std::endl;
-            correct_cfo(&samples[0], d_samples_per_symbol);
-        #endif
+            max = std::max_element(samples + gr::lora::clamp((int)(max - coeff),  0, (int)window),
+                                   samples + gr::lora::clamp(max +        coeff, 0u,      window)) - samples;
+            min = std::min_element(samples + gr::lora::clamp(max +           1u, 0u,      window),
+                                   samples + gr::lora::clamp(max +   2u * coeff, 0u,      window)) - samples;
 
-        samples_to_file("/tmp/data", &samples[0], d_samples_per_symbol, sizeof(gr_complex));
+            max += (min - max) / 2;
 
-        // Multiply with ideal downchirp
-        for(uint32_t i = 0; i < d_samples_per_symbol; i++) {
-            mult_hf[i] = conj(samples[i] * d_downchirp[i]);
-        }
+            *index = max;
 
-        samples_to_file("/tmp/mult", &mult_hf[0], d_samples_per_symbol, sizeof(gr_complex));
+            // Extra to allow falling edge on 99% of sample == Practically synced
+            // Should improve detection by not discarding almost synced samples, but doesn't..
+//            if ((max - coeff / 2) > (window * 0.98f)) {
+////                printf("ALREADY SYNCED\n");
+//                *index = 0;
+//                return 0.99f;
+//            }
 
-        // Perform decimation
-        for (uint32_t i = 0; i < d_number_of_bins; i++) {
-            firdecim_crcf_execute(d_decim, &mult_hf[d_decim_factor*i], &d_mult[i]);
-        }
+            float max_correlation = 0.0f;
 
-        samples_to_file("/tmp/resampled", &d_mult[0], d_number_of_bins, sizeof(gr_complex));
+            const float average = std::accumulate(samples + max, samples + window, 0.0f) / (window - max);
```

Comment:
This part is identical to the downchirp correlation, so in my opinion it's cleaner to create a general cross-correlation function where we can pass the average, standard deviation, and samples of the ideal down and upchirps.